<a href="https://colab.research.google.com/github/next-in-cmd/TASK02-Data-Cleaning-and-Processing-/blob/main/TASK02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing pandas library

In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv('/content/01.Data Cleaning and Preprocessing.csv')

In [ ]:
type(data)

pandas.core.frame.DataFrame

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 324 entries, 0 to 323
Data columns (total 23 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Observation      324 non-null    object 
 1   Y-Kappa          324 non-null    float64
 2   ChipRate         319 non-null    float64
 3   BF-CMratio       307 non-null    float64
 4   BlowFlow         308 non-null    float64
 5   ChipLevel4       323 non-null    float64
 6   T-upperExt-2     322 non-null    float64
 7   T-lowerExt-2     322 non-null    float64
 8   UCZAA            299 non-null    float64
 9   WhiteFlow-4      323 non-null    float64
 10  AAWhiteSt-4      173 non-null    float64
 11  AA-Wood-4        323 non-null    float64
 12  ChipMoisture-4   323 non-null    float64
 13  SteamFlow-4      323 non-null    float64
 14  Lower-HeatT-3    322 non-null    float64
 15  Upper-HeatT-3    322 non-null    float64
 16  ChipMass-4       323 non-null    float64
 17  WeakLiquorF     

Describing the data

In [ ]:
data.describe()

Y-Kappa    ChipRate  BF-CMratio     BlowFlow  ChipLevel4   \
count  301.000000  297.000000  287.000000   288.000000   300.000000   
mean    20.568605   14.338670   87.271564  1235.537278   259.365993   
std      2.990751    1.490121    7.839690   102.023065    85.643939   
min     12.170000    9.983000   68.645000     0.000000     0.000000   
25%     18.450000   13.358000   81.753500  1194.047250   215.369250   
50%     20.740000   14.417000   86.705000  1267.130000   271.845500   
75%     23.000000   15.492000   92.153500  1288.390000   321.285000   
max     27.600000   16.958000  121.717000  1351.240000   419.014000   

       T-upperExt-2   T-lowerExt-2         UCZAA  WhiteFlow-4   AAWhiteSt-4   \
count     300.000000      300.000000  277.000000    300.000000    160.000000   
mean      356.907033      324.010783    1.490588    593.148090      6.143012   
std         8.954457        7.650502    0.108138     66.949732      0.082396   
min       339.168000      284.633000    1.182000    405.111000      5.890000   
25%       350.317000      321.534000    1.429000    543.137750      6.093000   
50%       357.178500      325.638500    1.498000    594.294000      6.140000   
75%       362.326250      329.129250    1.561000    643.567000      6.200000   
max       399.135000      337.012000    1.747000    731.394000      6.340000   

       ...  SteamFlow-4   Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   \
count  ...    300.000000      300.00000      300.000000   300.000000   
mean   ...     66.834567      325.37041      300.381743   162.120610   
std    ...      5.643655        4.63735        4.647201    14.129462   
min    ...     48.568000      318.05100      293.312000   113.922000   
25%    ...     62.847250      321.23275      296.418250   153.089750   
50%    ...     67.751000      323.25150      298.493000   163.601000   
75%    ...     71.650250      329.81600      304.228250   172.286750   
max    ...     76.147000      333.85400      311.146000   189.268000   

       WeakLiquorF   BlackFlow-2   WeakWashF   SteamHeatF-3   T-Top-Chips-4   \
count    300.000000    300.000000  300.000000     300.000000      300.000000   
mean     875.491563   1170.387177  266.104633      49.664537      251.229687   
std      123.182165    149.288972  163.222997       4.589876        1.293151   
min      486.938000    838.948000    0.000000      35.510000      248.359000   
25%      791.096750   1038.338500  141.560000      46.378500      250.298250   
50%      869.552000   1144.513500  280.451500      50.192500      251.357500   
75%      969.904000   1318.632750  405.887750      53.284750      252.332250   
max     1226.277000   1395.767000  715.715000      63.332000      254.122000   

       SulphidityL-4   
count      160.000000  
mean        30.463594  
std          0.671066  
min         29.010000  
25%         30.107500  
50%         30.395000  
75%         30.838250  
max         32.840000  

[8 rows x 22 columns]

In [ ]:
data.tail(5)

Observation  Y-Kappa  ChipRate  BF-CMratio  BlowFlow  ChipLevel4   \
298    12-09:00    20.90    15.167      84.640  1283.706      339.440   
299    12-10:00    24.98       NaN      85.034  1278.345      368.564   
300    12-11:00    21.00       NaN      88.013  1307.722      278.842   
301    12-12:00    21.40       NaN      85.490  1255.986      273.484   
307    31-05:00    20.89    14.308      94.172  1327.832      251.120   

     T-upperExt-2   T-lowerExt-2    UCZAA  WhiteFlow-4   ...  SteamFlow-4   \
298        354.803         311.041  1.635       532.419  ...        65.561   
299        357.723         321.387    NaN       520.365  ...        65.729   
300        357.438         323.757    NaN       553.070  ...        65.795   
301        361.365         322.689    NaN       590.199  ...        71.456   
307        351.263         332.485  1.522       631.514  ...        71.286   

     Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   WeakLiquorF   BlackFlow-2   \
298        332.924         307.626      145.299       832.906      1344.708   
299        332.523         307.169      151.544       905.639      1344.469   
300        331.263         306.400      157.954       908.691      1344.588   
301        333.032         308.732      174.069       986.206      1348.747   
307        328.699         300.706      180.229       903.605      1323.082   

     WeakWashF   SteamHeatF-3   T-Top-Chips-4   SulphidityL-4   
298     388.911         49.524         251.833           30.29  
299     418.979         48.135         251.614           30.47  
300     462.712         54.373         251.197             NaN  
301     457.313         53.194         251.324           30.46  
307     232.729         54.503         250.084             NaN  

[5 rows x 23 columns]

 **Handling Values**

1. Drop_duplicates : Drop the duplicate values in the dataset.

In [ ]:
data = data.drop_duplicates()


2. Understanding Null Values

True : describes null value

False : describes no null value

In [ ]:
data.isnull()

Observation  Y-Kappa  ChipRate  BF-CMratio  BlowFlow  ChipLevel4   \
0          False    False     False       False     False        False   
1          False    False     False       False     False        False   
2          False    False     False       False     False        False   
3          False    False     False       False     False        False   
4          False    False     False       False     False        False   
..           ...      ...       ...         ...       ...          ...   
298        False    False     False       False     False        False   
299        False    False      True       False     False        False   
300        False    False      True       False     False        False   
301        False    False      True       False     False        False   
307        False    False     False       False     False        False   

     T-upperExt-2   T-lowerExt-2    UCZAA  WhiteFlow-4   ...  SteamFlow-4   \
0            False           False  False         False  ...         False   
1            False           False  False         False  ...         False   
2            False           False  False         False  ...         False   
3            False           False  False         False  ...         False   
4            False           False   True         False  ...         False   
..             ...             ...    ...           ...  ...           ...   
298          False           False  False         False  ...         False   
299          False           False   True         False  ...         False   
300          False           False   True         False  ...         False   
301          False           False   True         False  ...         False   
307          False           False  False         False  ...         False   

     Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   WeakLiquorF   BlackFlow-2   \
0            False           False        False         False         False   
1            False           False        False         False         False   
2            False           False        False         False         False   
3            False           False        False         False         False   
4            False           False        False         False         False   
..             ...             ...          ...           ...           ...   
298          False           False        False         False         False   
299          False           False        False         False         False   
300          False           False        False         False         False   
301          False           False        False         False         False   
307          False           False        False         False         False   

     WeakWashF   SteamHeatF-3   T-Top-Chips-4   SulphidityL-4   
0         False          False           False            True  
1         False          False           False           False  
2         False          False           False            True  
3         False          False           False           False  
4         False          False           False           False  
..          ...            ...             ...             ...  
298       False          False           False           False  
299       False          False           False           False  
300       False          False           False            True  
301       False          False           False           False  
307       False          False           False            True  

[301 rows x 23 columns]

In [ ]:
data.isnull().sum()

Observation          0
Y-Kappa              0
ChipRate             4
BF-CMratio          14
BlowFlow            13
ChipLevel4           1
T-upperExt-2         1
T-lowerExt-2         1
UCZAA               24
WhiteFlow-4          1
AAWhiteSt-4        141
AA-Wood-4            1
ChipMoisture-4       1
SteamFlow-4          1
Lower-HeatT-3        1
Upper-HeatT-3        1
ChipMass-4           1
WeakLiquorF          1
BlackFlow-2          1
WeakWashF            1
SteamHeatF-3         1
T-Top-Chips-4        1
SulphidityL-4      141
dtype: int64

In [ ]:
#It is just opposite to isnull function
#True : describes no null values
#False : describes nul values
data.notnull()


Observation  Y-Kappa  ChipRate  BF-CMratio  BlowFlow  ChipLevel4   \
0           True     True      True        True      True         True   
1           True     True      True        True      True         True   
2           True     True      True        True      True         True   
3           True     True      True        True      True         True   
4           True     True      True        True      True         True   
..           ...      ...       ...         ...       ...          ...   
298         True     True      True        True      True         True   
299         True     True     False        True      True         True   
300         True     True     False        True      True         True   
301         True     True     False        True      True         True   
307         True     True      True        True      True         True   

     T-upperExt-2   T-lowerExt-2    UCZAA  WhiteFlow-4   ...  SteamFlow-4   \
0             True            True   True          True  ...          True   
1             True            True   True          True  ...          True   
2             True            True   True          True  ...          True   
3             True            True   True          True  ...          True   
4             True            True  False          True  ...          True   
..             ...             ...    ...           ...  ...           ...   
298           True            True   True          True  ...          True   
299           True            True  False          True  ...          True   
300           True            True  False          True  ...          True   
301           True            True  False          True  ...          True   
307           True            True   True          True  ...          True   

     Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   WeakLiquorF   BlackFlow-2   \
0             True            True         True          True          True   
1             True            True         True          True          True   
2             True            True         True          True          True   
3             True            True         True          True          True   
4             True            True         True          True          True   
..             ...             ...          ...           ...           ...   
298           True            True         True          True          True   
299           True            True         True          True          True   
300           True            True         True          True          True   
301           True            True         True          True          True   
307           True            True         True          True          True   

     WeakWashF   SteamHeatF-3   T-Top-Chips-4   SulphidityL-4   
0          True           True            True           False  
1          True           True            True            True  
2          True           True            True           False  
3          True           True            True            True  
4          True           True            True            True  
..          ...            ...             ...             ...  
298        True           True            True            True  
299        True           True            True            True  
300        True           True            True           False  
301        True           True            True            True  
307        True           True            True           False  

[301 rows x 23 columns]

In [ ]:
# It Represent the total number of null values
data.isnull().sum().sum()

352

3.***Handling Null values***

1. Fillna()

In [ ]:
data2 = data.fillna(value=0)

In [ ]:
data2.tail(5)

Y-Kappa  ChipRate  BF-CMratio  BlowFlow  ChipLevel4   T-upperExt-2   \
276    22.70    15.517      83.008  1288.010      306.886        350.155   
296    20.50    13.358      97.662  1304.597      377.678        347.672   
297    20.40    14.233      89.790  1278.006      379.458        354.290   
298    20.90    15.167      84.640  1283.706      339.440        354.803   
307    20.89    14.308      94.172  1327.832      251.120        351.263   

     T-lowerExt-2    UCZAA  WhiteFlow-4   AAWhiteSt-4   ...  SteamFlow-4   \
276         322.485  1.590       568.752          6.17  ...        67.678   
296         313.147  1.546       496.460          6.34  ...        60.119   
297         315.558  1.515       491.374          0.00  ...        60.424   
298         311.041  1.635       532.419          6.34  ...        65.561   
307         332.485  1.522       631.514          0.00  ...        71.286   

     Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   WeakLiquorF   BlackFlow-2   \
276        331.854         309.346      160.061       910.013      1381.389   
296        332.615         308.575      141.076       997.904      1334.703   
297        331.980         308.078      140.301       975.016      1344.835   
298        332.924         307.626      145.299       832.906      1344.708   
307        328.699         300.706      180.229       903.605      1323.082   

     WeakWashF   SteamHeatF-3   T-Top-Chips-4   SulphidityL-4   
276     441.934         51.466         252.216           29.59  
296     389.497         46.206         252.423           30.43  
297     388.676         47.803         252.311            0.00  
298     388.911         49.524         251.833           30.29  
307     232.729         54.503         250.084            0.00  

[5 rows x 22 columns]

Row 299 to 301 & column chip rate shows the null values.

In [ ]:
data2.isnull().sum().sum()

0

2. Forward Filling

In [ ]:
data3 = data.fillna(method='pad')

<ipython-input-42-f14bc986d996>:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data3 = data.fillna(method='pad')


In [ ]:
data3.tail(5)

Observation  Y-Kappa  ChipRate  BF-CMratio  BlowFlow  ChipLevel4   \
298    12-09:00    20.90    15.167      84.640  1283.706      339.440   
299    12-10:00    24.98    15.167      85.034  1278.345      368.564   
300    12-11:00    21.00    15.167      88.013  1307.722      278.842   
301    12-12:00    21.40    15.167      85.490  1255.986      273.484   
307    31-05:00    20.89    14.308      94.172  1327.832      251.120   

     T-upperExt-2   T-lowerExt-2    UCZAA  WhiteFlow-4   ...  SteamFlow-4   \
298        354.803         311.041  1.635       532.419  ...        65.561   
299        357.723         321.387  1.635       520.365  ...        65.729   
300        357.438         323.757  1.635       553.070  ...        65.795   
301        361.365         322.689  1.635       590.199  ...        71.456   
307        351.263         332.485  1.522       631.514  ...        71.286   

     Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   WeakLiquorF   BlackFlow-2   \
298        332.924         307.626      145.299       832.906      1344.708   
299        332.523         307.169      151.544       905.639      1344.469   
300        331.263         306.400      157.954       908.691      1344.588   
301        333.032         308.732      174.069       986.206      1348.747   
307        328.699         300.706      180.229       903.605      1323.082   

     WeakWashF   SteamHeatF-3   T-Top-Chips-4   SulphidityL-4   
298     388.911         49.524         251.833           30.29  
299     418.979         48.135         251.614           30.47  
300     462.712         54.373         251.197           30.47  
301     457.313         53.194         251.324           30.46  
307     232.729         54.503         250.084           30.46  

[5 rows x 23 columns]

row from 298 to 301 & chip rate column shows forward filling results.

3. Backward Filling

In [ ]:
data4 = data.fillna(method='bfill')

<ipython-input-50-2b77f275dcd0>:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data4 = data.fillna(method='bfill')


In [ ]:
data4.tail(5)

Observation  Y-Kappa  ChipRate  BF-CMratio  BlowFlow  ChipLevel4   \
298    12-09:00    20.90    15.167      84.640  1283.706      339.440   
299    12-10:00    24.98    14.308      85.034  1278.345      368.564   
300    12-11:00    21.00    14.308      88.013  1307.722      278.842   
301    12-12:00    21.40    14.308      85.490  1255.986      273.484   
307    31-05:00    20.89    14.308      94.172  1327.832      251.120   

     T-upperExt-2   T-lowerExt-2    UCZAA  WhiteFlow-4   ...  SteamFlow-4   \
298        354.803         311.041  1.635       532.419  ...        65.561   
299        357.723         321.387  1.522       520.365  ...        65.729   
300        357.438         323.757  1.522       553.070  ...        65.795   
301        361.365         322.689  1.522       590.199  ...        71.456   
307        351.263         332.485  1.522       631.514  ...        71.286   

     Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   WeakLiquorF   BlackFlow-2   \
298        332.924         307.626      145.299       832.906      1344.708   
299        332.523         307.169      151.544       905.639      1344.469   
300        331.263         306.400      157.954       908.691      1344.588   
301        333.032         308.732      174.069       986.206      1348.747   
307        328.699         300.706      180.229       903.605      1323.082   

     WeakWashF   SteamHeatF-3   T-Top-Chips-4   SulphidityL-4   
298     388.911         49.524         251.833           30.29  
299     418.979         48.135         251.614           30.47  
300     462.712         54.373         251.197           30.46  
301     457.313         53.194         251.324           30.46  
307     232.729         54.503         250.084             NaN  

[5 rows x 23 columns]

Row from 299 to 307 & column Chip rate shows the result of back filling.

**Cleaning Process**

In [ ]:
import numpy as pandas
from scipy import stats

In [ ]:
#detect the outliers using IQR
data2.columns

Index(['Observation', 'Y-Kappa', 'ChipRate', 'BF-CMratio', 'BlowFlow',
       'ChipLevel4 ', 'T-upperExt-2 ', 'T-lowerExt-2  ', 'UCZAA',
       'WhiteFlow-4 ', 'AAWhiteSt-4 ', 'AA-Wood-4  ', 'ChipMoisture-4 ',
       'SteamFlow-4 ', 'Lower-HeatT-3', 'Upper-HeatT-3 ', 'ChipMass-4 ',
       'WeakLiquorF ', 'BlackFlow-2 ', 'WeakWashF ', 'SteamHeatF-3 ',
       'T-Top-Chips-4 ', 'SulphidityL-4 '],
      dtype='object')

In [ ]:
Q1 = data2.quantile(0.25)
Q3 = data2.quantile(0.75)
IQR = Q3 - Q1
print(IQR)

Y-Kappa              4.550
ChipRate             2.233
BF-CMratio          10.912
BlowFlow            96.766
ChipLevel4         105.868
T-upperExt-2        11.994
T-lowerExt-2         7.609
UCZAA                0.152
WhiteFlow-4        100.098
AAWhiteSt-4          6.143
AA-Wood-4            1.486
ChipMoisture-4       2.186
SteamFlow-4          8.840
Lower-HeatT-3        8.585
Upper-HeatT-3        7.852
ChipMass-4          19.347
WeakLiquorF        180.613
BlackFlow-2        280.829
WeakWashF          267.219
SteamHeatF-3         6.903
T-Top-Chips-4        2.044
SulphidityL-4       30.420
dtype: float64


In [ ]:
data2=data2[~((data2<(Q1-1.5*IQR))|(data2>(Q3+1.5*IQR))).any(axis=1)]

In [ ]:
data2.describe()

Y-Kappa    ChipRate  BF-CMratio     BlowFlow  ChipLevel4   \
count  226.000000  226.000000  226.000000   226.000000   226.000000   
mean    20.690487   14.673491   85.882181  1255.288916   264.664912   
std      2.982916    1.297369    7.033155    47.896055    74.345135   
min     12.480000   10.833000   68.645000  1084.083000    61.783000   
25%     18.457500   13.850000   80.984000  1221.926000   220.356000   
50%     20.775000   14.729000   84.967000  1280.291500   270.965000   
75%     23.010000   15.708000   91.178750  1289.254000   322.492000   
max     27.600000   16.958000  108.104000  1351.240000   419.014000   

       T-upperExt-2   T-lowerExt-2         UCZAA  WhiteFlow-4   AAWhiteSt-4   \
count     226.000000      226.000000  226.000000    226.000000    226.000000   
mean      356.861681      325.341124    1.487146    603.242482      3.098164   
std         7.466897        5.557537    0.108054     61.052197      3.078138   
min       340.222000      310.421000    1.182000    468.841000      0.000000   
25%       350.704250      322.355500    1.429000    549.611000      0.000000   
50%       357.560500      326.508500    1.492000    602.508000      5.904500   
75%       361.555000      329.264500    1.556000    653.358500      6.140000   
max       375.047000      337.012000    1.712000    731.394000      6.340000   

       ...  SteamFlow-4   Lower-HeatT-3  Upper-HeatT-3   ChipMass-4   \
count  ...    226.000000     226.000000      226.000000   226.000000   
mean   ...     67.545478     324.752212      299.655420   164.220102   
std    ...      4.914301       4.526481        4.383788    11.423269   
min    ...     52.962000     318.051000      293.312000   133.878000   
25%    ...     63.954000     321.179500      296.338500   156.091000   
50%    ...     68.147000     322.380000      297.636500   164.333000   
75%    ...     71.760750     329.575000      303.777000   172.555000   
max    ...     75.974000     333.223000      309.854000   189.268000   

       WeakLiquorF   BlackFlow-2   WeakWashF   SteamHeatF-3   T-Top-Chips-4   \
count    226.000000    226.000000  226.000000     226.000000      226.000000   
mean     874.123035   1149.895257  273.739403      49.810239      251.177779   
std      120.259977    150.321416  163.452307       4.143153        1.221296   
min      596.446000    838.948000    0.000000      38.283000      248.359000   
25%      784.366750   1014.977000  149.331750      46.639750      250.290750   
50%      866.170000   1126.513500  283.079500      50.128500      251.233000   
75%      968.683250   1302.847000  414.599750      52.889250      252.240000   
max     1132.181000   1392.868000  715.715000      59.564000      254.122000   

       SulphidityL-4   
count      226.000000  
mean        15.391987  
std         15.297984  
min          0.000000  
25%          0.000000  
50%         29.065000  
75%         30.437500  
max         32.840000  

[8 rows x 22 columns]